### PR TITLE
Towards #12251 part 2, replace all full(X) calls with convert(Array, X) and migrate tests

### PR DESCRIPTION
**Pull request by Sacha0**
Second step towards JuliaLang/LinearAlgebra.jl#231 (for a description of the overall objective and plan, see https://github.com/JuliaLang/julia/pull/17066#issuecomment-228113687). This pull request replaces all `full(X)` calls with `convert(Array, X)` and accordingly migrates tests and documentation. Thanks and best!

### CODE REVIEW

File: doc/manual/arrays.rst [hunk 1]
```diff
@@ -811,7 +811,7 @@ reference.
 +----------------------------------------+----------------------------------+--------------------------------------------+
 | :func:`speye(n) <speye>`               | :func:`eye(n) <eye>`             | Creates a *n*-by-*n* identity matrix.      |
 +----------------------------------------+----------------------------------+--------------------------------------------+
-| :func:`full(S) <full>`                 | :func:`sparse(A) <sparse>`       | Interconverts between dense                |
+| :func:`convert(Array, S) <convert>`                 | :func:`sparse(A) <sparse>`       | Interconverts between dense                |
```
Comment (tkelman) [inline doc/manual/arrays.rst]: table alignment needs fixing

Comment (Sacha0) [inline doc/manual/arrays.rst]: Fixed. Thanks!

**Comment by andreasnoack** [comment]
@Sacha0 Thanks for all the work on this. With #17066 merged, I guess it's time for a rebase here.

**Comment by tkelman** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

File: test/sparsedir/sparse.jl [hunk 2]
```diff
@@ -499,42 +499,42 @@ for (aa116, ss116) in [(a116, s116), (ad116, sd116)]
...
+    @test convert(Array, ss116[end:-2:1,j]) == aa116[end:-2:1,j]
     # float-range indexing is not supported
 
     # sorted vector indexing
-    @test full(ss116[i,[3:2:end-3;]]) == aa116[i,[3:2:end-3;]]
-    @test full(ss116[[3:2:end-3;],j]) == aa116[[3:2:end-3;],j]
-    @test full(ss116[i,[end-3:-2:1;]]) == aa116[i,[end-3:-2:1;]]
-    @test full(ss116[[end-3:-2:1;],j]) == aa116[[end-3:-2:1;],j]
+    @test convert(Array, ss116[i,[3:2:end-3;]]) == aa116[i,[3:2:end-3;]]
+    @test convert(Array, ss116[[3:2:end-3;],j]) == aa116[[3:2:end-3;],j]
```
Comment (tkelman) [inline test/sparsedir/sparse.jl]: It's a bad sign when your diffs get long enough that github gives up on coloring words any more. I should probably have looked at this commit-by-commit.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17079#issuecomment-233127309) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/73be1338f5813fab7fdc51b479eceb0e5409956a/2981f30_vs_1fd440e/report.md). cc @jrevels

File: test/sparsedir/sparse.jl (hunk 2 above)
Comment (Sacha0) [inline test/sparsedir/sparse.jl]: Apologies. How should I improve this? Thanks!

Comment (tkelman) [inline test/sparsedir/sparse.jl]: no action needed, what you did with splitting to multiple commits was good, I just should have reviewed in the same way. or we can complain to github to highlight more of long diffs

**Comment by Sacha0** [comment]
Travis i686 timeout unrelated?

**Comment by Sacha0** [comment]
Thanks for the review and merge!

**Comment by tkelman** [comment]
This looks like it had some unintended consequences, ref https://github.com/JuliaStats/DataArrays.jl/issues/208 - not sure what that was doing before, I guess specializing `full` to return something other than `Array` ?

**Comment by tkelman** [comment]
Since this isn't really critical and wasn't supposed to have visible consequences, I think we may want to revert it for now and revisit after branching 0.5?

**Comment by Sacha0** [comment]
> Since this isn't really critical and wasn't supposed to have visible consequences, I think we may want to revert it for now and revisit after branching 0.5?

Sounds good! Prior to merger I assumed this was 0.6-dev material. Taking a look now at the base/abstractarray.jl code `@andreasnoack` identified (independent of reversion) and will report back. Thanks!

**Comment by Sacha0** [comment]
As `@andreasnoack` noted above, `full` in the generic concatenation methods' `similar(full(X), T, dims)` calls seems intended to handle annotation and special matrix types such as `Diagonal` and `LowerTriangular` and otherwise be a no-op. Specifically, when `X` is such an annotation or special matrix type, the `full` call returns a form of `X` that enables `similar(full(X), T, dims)` to return something concatenable yet, if possible, more similar to `X` than `Array`. On the other hand, the presence of `full` seems to make the generic concatenation methods inapplicable to sparse matrices/vectors and perhaps similar types.

One path forward would be to introduce `backingtype` and change the identified calls to `similar(convert(backingtype(X), X), T, dims)`. But `backingtype` would need to exist for a broad swathe of types, practically requiring introduction of a no-op `backingtype` fallback. And that no-op `backingtype` fallback would essentially be a reintroduction of `full`'s no-op fallback with a different name and slightly clearer definition, though still fraught with issues. (E.g., what should `backingtype` return for special matrix types? Need you abstract `backingtype` to something that provides the closest type that can have arbitrary shape and is fully writeable to make `similar(convert(backingtype(X), X), T, dims)` work as needed here? That sounds like a more principled replacement for `full`, but also sounds suspiciously like one conception of `similar`.) Moreover, in some sense this path would couple the logic of annotation and special matrix types with many unrelated things, whereas reducing that coupling was part of the appeal of deprecating `full`.

Another path forward would be removing `full` from the identified `similar(full(X), T, dims)` calls. This approach has the appeal of decoupling annotation and special matrix types' logic from the rest of the codebase. In the near term this approach would reduce the degree to which the generic concatenation methods preserve structure. But writing specialized `similar` methods could restore much of that ability, and specialized concatenation methods (as presently exist for e.g. sparse matrices/vectors due to the `full`s in question) could take care of the rest. This approach would make clarifying the meaning of `similar` (and consequently `convert` etc) more urgent, but that might be a good thing.

Thoughts? Did the `full` calls serve any other purpose? Best!

**Pull request merged**